src/caffe/net.cpp

```diff
@@ -104,7 +104,8 @@ void Net<Dtype>::Init(const NetParameter& in_param) {
     for (int top_id = 0; top_id < num_top; ++top_id) {
       AppendTop(param, layer_id, top_id, &available_blobs, &blob_name_to_idx);
       // Collect Input layer tops as Net inputs.
-      if (layer_param.type() == "Input") {
+      if (layer_param.type() == "Input" ||
+          layer_param.type().find("Data") != std::string::npos) {
```

Comment:
This check seems fragile.  Wouldn't a more general definition of input be any layer that has a top but no bottom?